### PR TITLE
Synchronous write to tag views table

### PR DESCRIPTION
**Pull request by chbatey**
#259 

This changes the tag write to be part of the Future returned from asyncWriteMessages

It is still asynchronous with respect to the write to the messages table so it can still fail/timeout while the messages table write succeed however now the actor will be stopped. The tag_view will then be fixed when it is restarted.

Note that the TagWriter does keep retrying the write until the persistent actor stops, need to decide what to set that timeout to.

### CODE REVIEW

**Comment by chbatey** [comment]
Found some issues during load testing so hold off reviewing this one for now

**Review by patriknw — COMMENTED** [review]
Looked briefly at this and seems like a good approach. I will try to review more thoroughly next week. It's a substantial change but still not as scary as I expected.

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala [hunk 1]
```diff
@@ -233,15 +365,16 @@ import akka.util.UUIDComparator
       awaitingFlush match {
         case Some(replyTo) =>
           log.debug("External flush request")
-          if (sortedBuffer.nonEmpty) {
-            // TODO break into batches
-            write(sortedBuffer, Vector.empty[(Serialized, TagPidSequenceNr)], tagPidSequenceNrs, awaitingFlush)
+          if (buffer.pending.nonEmpty) {
+            // TODO break into batches - check, but this is now done as buffer always batches
```
Comment (patriknw) [inline core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala:369]: still TODO?

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala [hunk 2]
```diff
@@ -233,15 +365,16 @@ import akka.util.UUIDComparator
...
+            write(nextBuffer, tagPidSequenceNrs, awaitingFlush)
           } else {
             replyTo ! FlushComplete
-            context.become(idle(sortedBuffer, tagPidSequenceNrs))
+            context.become(idle(nextBuffer, tagPidSequenceNrs))
           }
         case None =>
-          flushIfRequired(sortedBuffer, tagPidSequenceNrs)
+          log.debug("write finished, checking if a new flush is required")
```
Comment (patriknw) [inline core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala:376]: is this logged for each write? too much? (do we have a verbose logging flag here?)

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala [hunk 3]
```diff
@@ -253,15 +386,11 @@ import akka.util.UUIDComparator
         t)
       timers.startSingleTimer(FlushKey, InternalFlush, settings.flushInterval)
       parent ! TagWriters.TagWriteFailed(t)
-      context.become(idle(events ++ buffer, tagPidSequenceNrs))
+      context.become(idle(buffer, tagPidSequenceNrs))
```
Comment (patriknw) [inline core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala:389]: does the buffer still contain the `events` or is this that we don't retry them in this way any more?

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala [hunk 4]
```diff
@@ -278,36 +407,17 @@ import akka.util.UUIDComparator
...
-        write(buffer.take(settings.maxBatchSize), buffer.drop(settings.maxBatchSize), tagSequenceNrs, None)
-      } else {
-        write(currentBucket, rest, tagSequenceNrs, None)
-      }
-    } else if (buffer.size >= settings.maxBatchSize) {
-      log.debug("Batch size reached. Writing to Cassandra.")
-      write(buffer.take(settings.maxBatchSize), buffer.drop(settings.maxBatchSize), tagSequenceNrs, None)
-    } else if (settings.flushInterval == Duration.Zero) {
-      // Should always be a buffer of 1
-      write(buffer, Vector.empty[(Serialized, TagPidSequenceNr)], tagSequenceNrs, None)
```
Comment (patriknw) [inline core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala:305]: nice

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala [hunk 5]
```diff
@@ -329,42 +439,46 @@ import akka.util.UUIDComparator
...
-      }
-    })
+  private def createTagWriteSummary(writes: Buffer): Map[PersistenceId, PidProgress] = {
+    writes.nextBatch
+      .flatten(_.events)
+      .foldLeft(Map.empty[PersistenceId, PidProgress])((acc, next) => {
+        val (event, tagPidSequenceNr) = next
+        acc.get(event.persistenceId) match {
+          case Some(PidProgress(from, to, _, _)) =>
+            // FIXME make this log nicer
```
Comment (patriknw) [inline core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala:466]: ^ ?

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriters.scala [hunk 6]
```diff
@@ -209,9 +220,12 @@ import scala.util.Try
     case TagFlush(tag) =>
       tellTagActor(tag, Flush, sender())
     case tw: TagWrite =>
-      forwardTagWrite(tw)
+      // this only comes from the replay, an ack is not required right now.
+      val toReply = sender()
+      forwardTagWrite(tw).map(_ => Done).pipeTo(toReply)
```
Comment (patriknw) [inline core/src/main/scala/akka/persistence/cassandra/journal/TagWriters.scala:225]: `pipeTo(sender())` would be ok

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriters.scala [hunk 7]
```diff
@@ -209,9 +220,12 @@ import scala.util.Try
...
     case BulkTagWrite(tws, withoutTags) =>
-      tws.foreach(forwardTagWrite)
+      val toReply = sender()
+      Future.traverse(tws)(tw => forwardTagWrite(tw)).map(_ => Done).pipeTo(toReply)
```
Comment (patriknw) [inline core/src/main/scala/akka/persistence/cassandra/journal/TagWriters.scala:228]: same here

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriters.scala [hunk 8]
```diff
@@ -329,14 +343,15 @@ import scala.util.Try
...
-  private def forwardTagWrite(tw: TagWrite): Unit = {
+  private def forwardTagWrite(tw: TagWrite): Future[Done] = {
     if (tw.actorRunning && !currentPersistentActors.contains(tw.serialised.head.persistenceId)) {
       log.warning(
         "received TagWrite but actor not active (dropping, will be resolved when actor restarts): [{}]",
         tw.serialised.head.persistenceId)
+      Future.successful(Done)
     } else {
       updatePendingScanning(tw.serialised)
-      tellTagActor(tw.tag, tw, sender())
```
Comment (patriknw) [inline core/src/main/scala/akka/persistence/cassandra/journal/TagWriters.scala:339]: do we still need `tellTagActor` or can be simplify by changing to askTagActor for the remaining usage of tellTagActor?

File: core/src/test/scala/akka/persistence/cassandra/journal/TagWritersSpec.scala [hunk 9]
```diff
@@ -19,6 +19,7 @@ import org.scalatest.wordspec.AnyWordSpecLike
 import org.scalatest.matchers.should.Matchers
 import scala.concurrent.duration._
 
+// FIXME test for replies
```
Comment (patriknw) [inline core/src/test/scala/akka/persistence/cassandra/journal/TagWritersSpec.scala:22]: ^ ?

**Review by patriknw — COMMENTED** [review]
looking very good

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala [hunk 10]
```diff
@@ -83,15 +85,136 @@ import akka.util.UUIDComparator
...
+  /**
+   * Buffered events waiting to be written.
+   * The next batch is maintained in `nextBatch` and will never contain more than the `batchSize`
+   * or events from different time buckets.
+   *
+   * Events should be added and then call shouldWrite() to see if a batch is ready to be written.
+   * Once a write is complete call `writeComplete` to discard the events in `nextBatch` and take
+   * events from `pending` for the next batch.
+   */
+  case class Buffer(
```
[the remark this answers is not in the document]
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala:114]: 👍

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala [hunk 11]
```diff
@@ -196,21 +319,30 @@ import akka.util.UUIDComparator
...
           "If events are buffered for longer than the eventual-consistency-delay they won't be picked up by live queries. The oldest event in the buffer is offset: {}",
           buffer.size,
-          formatOffset(buffer.head._1.timeUuid))
+          formatOffset(buffer.nextBatch.head.events.head._1.timeUuid))
       }
       // buffer until current query is finished
       // Don't sort until the write has finished
-      become(writeInProgress(buffer ++ events, updatedTagPidSequenceNrs, awaitingFlush))
+      val newBuffer = buffer.addPending(awaitingWrite)
+      log.debug("buffering write as write in progress. New buffer {}", newBuffer)
```
[the remark this answers is not in the document]
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala:335]: Ya definitely. Removing this was just from when i was debugging

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala [hunk 12]
```diff
@@ -182,11 +304,12 @@ import akka.util.UUIDComparator
   }
 
   private def writeInProgress(
-      buffer: Vector[(Serialized, TagPidSequenceNr)],
+      buffer: Buffer,
       tagPidSequenceNrs: Map[PersistenceId, TagPidSequenceNr],
       awaitingFlush: Option[ActorRef]): Receive = {
     case DropState(pid) =>
       log.debug("Dropping state for pid: [{}]", pid)
+      // FIXME remove anything from the buffer
```
[the remark this answers is not in the document]
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala:312]: Done

**Comment by chbatey** [comment]
Also refs https://github.com/akka/akka-persistence-cassandra/issues/405

File: core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala [hunk 13]
```diff
@@ -280,9 +282,15 @@ import akka.stream.scaladsl.Source
 
             rec(groups)
           }
-        result.map { _ =>
-          tagWrites.foreach(_ ! extractTagWrites(serialized))
-          Nil
+
+        // TODO base this off some other timeout or put in config
+        // TODO make the error message helpful
```
[the remark this answers is not in the document]
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala:287]: Doing now

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala (hunk 1 above)
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala:369]: The places we didn't batch correctly are now both fixed as part of this PR

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala (hunk 3 above)
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala:389]: The buffer still contains the events, they are only removed once a writeComplete happens

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriters.scala (hunk 8 above)
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/journal/TagWriters.scala:339]: Nope! Seems we don't ever send a TagFlush to TagWriters anyhow

File: core/src/main/resources/reference.conf [hunk 14]
```diff
@@ -252,9 +252,10 @@ akka.persistence.cassandra {
     # Max time to buffer events for before writing.
     # Larger values will increase cassandra write efficiency but increase the delay before
     # seeing events in EventsByTag queries.
-    # Setting this to 0 means that tag writes will be written immediately but will still be asynchronous
-    # with respect to the PersistentActor's persist call.
-    flush-interval = 250ms
+    # Setting this to 0 means that tag writes will be written immediately. Batching will still happen
+    # as events are buffered while a write is in progress
+    # If the tag write fails for an event the actor's persist will be failed
```
[the remark this answers is not in the document]
Comment (chbatey) [reply core/src/main/resources/reference.conf]: I've added the explanation to the new timeout setting

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala (hunk 2 above)
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala:254]: Removed, was when debugging

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriters.scala [hunk 15]
```diff
@@ -329,14 +337,15 @@ import scala.util.Try
...
+      askTagActor(tw.tag, tw).map(_ => Done)
```
Comment (johanandren) [inline core/src/main/scala/akka/persistence/cassandra/journal/TagWriters.scala:348]: Another map that can be on parasitic ec

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriters.scala [hunk 16]
```diff
@@ -329,14 +337,15 @@ import scala.util.Try
 
   }
 
-  private def forwardTagWrite(tw: TagWrite): Unit = {
+  private def forwardTagWrite(tw: TagWrite): Future[Done] = {
```
Comment (johanandren) [inline core/src/main/scala/akka/persistence/cassandra/journal/TagWriters.scala:340]: This is called from a future callback but uses the actor logger and may call updatePendingScanning which mutates the actor state.

Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/journal/TagWriters.scala:340]: Good catch!

File: core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala [hunk 17]
```diff
@@ -280,9 +279,15 @@ import akka.stream.scaladsl.Source
 
             rec(groups)
           }
-        result.map { _ =>
-          tagWrites.foreach(_ ! extractTagWrites(serialized))
-          Nil
+
+        // The tag writer keeps retrying but will drop writes for a persistent actor when it restarts
+        // due to this failing
+        implicit val timeout: Timeout = Timeout(settings.eventsByTagSettings.tagWriteTimeout)
```
[the remark this answers is not in the document]
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala:285]: i'll move it inside

File: core/src/main/scala/akka/persistence/cassandra/journal/TagWriters.scala (hunk 16 above)
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/journal/TagWriters.scala:340]: Actually, the traverse calls this on the calling thread but i think doing that explicitly is more clear

**Comment by chbatey** [comment]
@patriknw do you mind taking one last look at this before we merge?

**Review by patriknw — APPROVED** [review]
LGTM (I looked at the additional review commits)

**Review by johanandren — APPROVED** [review]

File: core/src/main/resources/reference.conf [hunk 18]
```diff
@@ -252,9 +252,18 @@ akka.persistence.cassandra {
     # Max time to buffer events for before writing.
     # Larger values will increase cassandra write efficiency but increase the delay before
     # seeing events in EventsByTag queries.
-    # Setting this to 0 means that tag writes will be written immediately but will still be asynchronous
-    # with respect to the PersistentActor's persist call.
-    flush-interval = 250ms
+    # Setting this to 0 means that tag writes will be written immediately. Batching will still happen
```
Comment (agapic) [inline core/src/main/resources/reference.conf:255]: Just came across this PR when debugging an issue: https://discuss.lightbend.com/t/akka-persistence-cassandra-unrecoverable-failures-when-tag-writes-buffer-gets-too-big/7643

We're using version 1.0.1 of this, would upgrading to 1.0.4 help resolve the issue we're seeing?

Comment (chbatey) [reply core/src/main/resources/reference.conf:255]: Replied on the thread

**Pull request merged**